python/singa/autograd.py

```diff
@@ -358,6 +358,51 @@ def relu(x):
     return ReLU()(x)[0]
 
 
+class HardSigmoid(Operation):
+    def __init__(self,alpha=0.2,gamma=0.5):
+        super(HardSigmoid, self).__init__()
+        self.alpha=alpha
+        self.gamma=gamma
+
+    def forward(self, x):
+        """Do forward propgation.
+        #y = max(0, min(1, alpha * x + gamma))
+        Args:
+            x (CTensor): matrix
+        Returns:
+            a CTensor for the result
+        """
+        if training:
+            self.input = x
+
+        x = singa.AddFloat(singa.MultFloat(x,self.alpha),self.gamma)
+        x = singa.ReLU(x)
+        mask = singa.LTFloat(x, 1.0)
+        mask2 = singa.GEFloat(x, 1.0)
+
+        ans = singa.__add__(singa.__mul__(x, mask),mask2)
```

Comment:
why add the mask2 here?

Reply:
if x >=1 then x == 1
I first made x[x>=1]=0 and then x[x>=1]+=1

Reply:
can we write like this?
```
x = singa.AddFloat(singa.MultFloat(x, self.alpha), self.gamma)
x = singa.__clamp__(x, 0.0, 1.0)
```

Reply:
relu will turn x into x if x>=0 and 0 otherwise;
then, you do not need to add 1?